File: ruby/bob/bob.rb
class Bob
  class << self
    def hey(remark)
      remark= remark.strip
      return 'Fine. Be that way!' if remark.empty?
      return "Calm down, I know what I'm doing!" if shouting?(remark) && question?(remark)
      return 'Whoa, chill out!' if shouting?(remark)
      return 'Sure.' if question?(remark)
      'Whatever.'
    end

    private

    def shouting?(remark)
      remark.upcase == remark && remark.match(/[A-Z]/)
    end

    def question?(remark)
      remark.end_with?('?')
    end
  end
end

File: elixir/bob/bob.exs
import String

defmodule Bob do
  def hey(input) do
    cond do
      silence?(input)           -> "Fine. Be that way!"
      forceful_question?(input) -> "Calm down, I know what I'm doing!"
      question?(input)          -> "Sure."
      shouting?(input)          -> "Whoa, chill out!"
      true                      -> "Whatever."
    end
  end

  defp silence?(input), do: trim(input) == ""

  defp forceful_question?(input), do: question?(input) && shouting?(input)

  defp question?(input), do: input |> ends_with?("?")

  defp shouting?(input), do: upcase(input) == input && upcase(input) != downcase(input)
end
